// @apollo/query-planner/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryPlanner = exports.prettyFormatQueryPlan = exports.astSerializer = exports.queryPlanSerializer = void 0;
var snapshotSerializers_1 = require("./snapshotSerializers");
Object.defineProperty(exports, "queryPlanSerializer", { enumerable: true, get: function () { return snapshotSerializers_1.queryPlanSerializer; } });
Object.defineProperty(exports, "astSerializer", { enumerable: true, get: function () { return snapshotSerializers_1.astSerializer; } });
var prettyFormatQueryPlan_1 = require("./prettyFormatQueryPlan");
Object.defineProperty(exports, "prettyFormatQueryPlan", { enumerable: true, get: function () { return prettyFormatQueryPlan_1.prettyFormatQueryPlan; } });
__exportStar(require("./QueryPlan"), exports);
const query_graphs_1 = require("@apollo/query-graphs");
const buildPlan_1 = require("./buildPlan");
class QueryPlanner {
    constructor(supergraphSchema) {
        this.supergraphSchema = supergraphSchema;
        this.federatedQueryGraph = (0, query_graphs_1.buildFederatedQueryGraph)(supergraphSchema, true);
    }
    buildQueryPlan(operation) {
        if (operation.selectionSet.isEmpty()) {
            return { kind: 'QueryPlan' };
        }
        return (0, buildPlan_1.computeQueryPlan)(this.supergraphSchema, this.federatedQueryGraph, operation);
    }
}
exports.QueryPlanner = QueryPlanner;
