// @protoplasm/recall/dist/recall.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.recall = void 0;
const report_1 = require("./report");
const weakish_trie_1 = __importDefault(require("./weakish-trie"));
function createRecall(fn, cache = weakish_trie_1.default.weakish()) {
    function call(...args) {
        const result = getResult.apply(this, args);
        (0, report_1.report)(result.log);
        return result.unwrap();
    }
    function getResult(...args) {
        const entry = cache.entry(this, ...args);
        const result = entry.value ? entry.value : entry.set((0, report_1.execute)(fn, this, args));
        return result;
    }
    function getExisting(...args) {
        return cache.entry(this, ...args).value;
    }
    function* eachExisting(...args) {
        const entry = cache.entry(this, ...args);
        yield* entry.entries();
    }
    call.getResult = getResult;
    call.getExisting = getExisting;
    call.eachExisting = eachExisting;
    return call;
}
exports.recall = createRecall;
exports.default = exports.recall;
