// @apollo/core-schema/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.err = exports.report = exports.getResult = exports.gql = exports.toDefinitionKind = exports.byKind = exports.byName = exports.hasName = exports.isAst = exports.flat = exports.first = exports.maybe = exports.maybeOne = exports.only = exports.groupBy = exports.directives = exports.schemaDefinitions = exports.byGref = exports.GRef = exports.Version = exports.LinkUrl = exports.Atlas = exports.Schema = exports.default = void 0;
var schema_1 = require("./schema");
Object.defineProperty(exports, "default", { enumerable: true, get: function () { return schema_1.Schema; } });
var schema_2 = require("./schema");
Object.defineProperty(exports, "Schema", { enumerable: true, get: function () { return schema_2.Schema; } });
var atlas_1 = require("./atlas");
Object.defineProperty(exports, "Atlas", { enumerable: true, get: function () { return atlas_1.Atlas; } });
var link_url_1 = require("./link-url");
Object.defineProperty(exports, "LinkUrl", { enumerable: true, get: function () { return link_url_1.LinkUrl; } });
var version_1 = require("./version");
Object.defineProperty(exports, "Version", { enumerable: true, get: function () { return version_1.Version; } });
var gref_1 = require("./gref");
Object.defineProperty(exports, "GRef", { enumerable: true, get: function () { return gref_1.GRef; } });
Object.defineProperty(exports, "byGref", { enumerable: true, get: function () { return gref_1.byGref; } });
var directives_1 = require("./directives");
Object.defineProperty(exports, "schemaDefinitions", { enumerable: true, get: function () { return directives_1.schemaDefinitions; } });
Object.defineProperty(exports, "directives", { enumerable: true, get: function () { return directives_1.directives; } });
var each_1 = require("./each");
Object.defineProperty(exports, "groupBy", { enumerable: true, get: function () { return each_1.groupBy; } });
Object.defineProperty(exports, "only", { enumerable: true, get: function () { return each_1.only; } });
Object.defineProperty(exports, "maybeOne", { enumerable: true, get: function () { return each_1.maybeOne; } });
Object.defineProperty(exports, "maybe", { enumerable: true, get: function () { return each_1.maybe; } });
Object.defineProperty(exports, "first", { enumerable: true, get: function () { return each_1.first; } });
Object.defineProperty(exports, "flat", { enumerable: true, get: function () { return each_1.flat; } });
var is_1 = require("./is");
Object.defineProperty(exports, "isAst", { enumerable: true, get: function () { return is_1.isAst; } });
Object.defineProperty(exports, "hasName", { enumerable: true, get: function () { return is_1.hasName; } });
Object.defineProperty(exports, "byName", { enumerable: true, get: function () { return is_1.byName; } });
Object.defineProperty(exports, "byKind", { enumerable: true, get: function () { return is_1.byKind; } });
Object.defineProperty(exports, "toDefinitionKind", { enumerable: true, get: function () { return is_1.toDefinitionKind; } });
var gql_1 = require("./gql");
Object.defineProperty(exports, "gql", { enumerable: true, get: function () { return gql_1.gql; } });
var recall_1 = require("@protoplasm/recall");
Object.defineProperty(exports, "getResult", { enumerable: true, get: function () { return recall_1.getResult; } });
Object.defineProperty(exports, "report", { enumerable: true, get: function () { return recall_1.report; } });
var error_1 = require("./error");
Object.defineProperty(exports, "err", { enumerable: true, get: function () { return error_1.err; } });
__exportStar(require("./errors"), exports);
